Button: deprecate `isPrimary`, `isSecondary`, `isTertiary` and `isLink` props in favour of `variant` prop

ciampo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

This PR takes inspiration from previous PRs  #29990 and  [#31297](https://github.com/WordPress/gutenberg/pull/31297/files#diff-4b2d1f303d1a1bc73ea0905ebdad1ed48faa9e800d68aa1c265c9287ff30efe5R25-R38)

- Expose a new `variant` prop (i.e. `variant="primary|secondary|tertiary|link"`)
- Do not mark the `isPrimary`, `isSecondary`, `isTertiary` and `isLink` props as deprecated yet, but add logic to the `<Button />` component to _translate_ their values into the `variant` prop
- Update docs, changelog, unit tests and Storybook Button stories
- Add unit tests to verify that the above-mentioned props are effectively marked as deprecated, while the expected classes are still being applied
- Update all of the usages in the repo to utilise the new `variant` prop instead of the deprecated props (necessary to have all tests passing)

Part of the next steps highlighted in #30503

## Testing instructions
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

- [ ] Run Gutenberg and make sure that Buttons still work as expected
- [ ] All tests are passing
- [ ] Run Button unit tests: `npm run test-unit packages/components/src/button`
- [ ] Run local storybook `npm run storybook:dev` and compare the Button component against what's currently in [production](https://wordpress.github.io/gutenberg/?path=/story/components-button--buttons)
- [ ] Search in the repo's codebase for the `isPrimary`, `isSecondary`, `isTertiary` and `isLink` props and make sure they have all been replaced with the new `variant` prop

## Screenshots <!-- if applicable -->

N/A

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] N/A I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ciampo committed:
Deprecate isPrimary, isSecondary and isTertiary props in favor of variant prop

ciampo committed:
Update button docs

ciampo committed:
Update Button variant in Storybook

ciampo committed:
Update tests

ciampo committed:
changelog

ciampo committed:
Swap order of isSecondary / isTertiary deprecation

ciampo committed:
Depracate isLink prop too

ciampo committed:
Spacing

ciampo commented on packages/components/src/button/index.js:
```diff
@@ -19,18 +19,76 @@ import { VisuallyHidden } from '../visually-hidden';
+		computedVariant ??= 'secondary';
+	}
+
+	if ( isDefault ) {
+		deprecated( 'Button isDefault prop', {
+			since: '5.4',
+			alternative: 'variant="secondary"',
+		} );
+
+		computedVariant ??= 'secondary';
+	}
+
+	if ( isLink ) {
+		deprecated( 'Button isLink prop', {
+			since: '5.9',
+			alternative: 'variant="link"',
+		} );
+
+		computedVariant ??= 'link';
+	}
```
[This previous PR](https://github.com/WordPress/gutenberg/pull/29990/files#diff-e931867515d855b923dc4148f4e2f9277bf7b37cd766f95e28ac11df4eb0b375R10-R54) from @sarayourfriend was mapping `isDefault` to `plain`, while currently in production, it seems like [this prop is mapped to `secondary`](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/button/index.js#L59)

Sara's PR also uses `plain` as the fallback value — is it something that should be added to this PR too?

ciampo replied:
We decided to keep the code as-is currently — `isDefault` mapped to `secondary` and `undefined` as default / fallback value

ciampo committed:
Update changelog

ciampo committed:
Update example in README

ciampo commented on packages/components/src/button/index.js:
```diff
@@ -19,18 +19,76 @@ import { VisuallyHidden } from '../visually-hidden';
 
 const disabledEventsOnDisabledButton = [ 'onMouseDown', 'onClick' ];
 
+// TODO: add JSDOCS
```
Not sure how to add JS docs here. The example I was referring to for [Flex](https://github.com/WordPress/gutenberg/pull/31297/files#diff-4b2d1f303d1a1bc73ea0905ebdad1ed48faa9e800d68aa1c265c9287ff30efe5R20-R24) imports TypeScript types, which don't seem to be available for the `<Button />` component

ciampo replied:
As agreed, no need to add comments just yet

ciampo commented on packages/components/src/button/README.md:
```diff
@@ -137,29 +137,13 @@ If provided, renders `a` instead of `button`.
 -   Default: `undefined`
 
-#### isSecondary
+#### variant
```
I've updated the docs myself, not sure if this the correct way of doing so

sarayourfriend replied:
It is as there's no automatic documentation generated for un-annotated functions like `Button`.

ciampo committed:
Add unit test for deprecated props

kevin940726 commented on packages/components/src/button/test/deprecated-props.js:
```diff
@@ -0,0 +1,98 @@
+ */
+import { Button } from '..';
+
+// The full module mock is necessary to avoid test failures related to
+// calling `console` methods
+jest.mock( '@wordpress/deprecated', () => {
+	return {
+		__esModule: true,
+		default: jest.fn(),
+	};
+} );
```
We can use `expect(console).toHaveWarned(...)` instead of mocking, it should also be more semantic too IMO.

ciampo replied:
TIL about `@wordpress/jest-console` ! Thank you for the advice

Done in https://github.com/WordPress/gutenberg/pull/31713/commits/69ed9efd0c6674c8e7d627dc1971e875081bd4fa

kevin940726 commented on packages/components/src/button/test/deprecated-props.js:
```diff
@@ -0,0 +1,98 @@
+/**
```
I think this file can just be a test case in main `test/index.js` file. Don't need to create another test file.

ciampo replied:
The reason I created a separate file was to be able to mock the `@wordpress/deprecate` package without affecting the other `<Button />` tests. But thanks to your suggestion to use `expect( console ).toHaveWarned()`, that's not necessary anymore.

I went ahead and moved the tests back into `test/index.js` in https://github.com/WordPress/gutenberg/pull/31713/commits/69ed9efd0c6674c8e7d627dc1971e875081bd4fa

kevin940726 commented on packages/components/src/placeholder/style.scss:
```diff
@@ -1,4 +1,5 @@
-.components-placeholder.components-placeholder { // This needs specificity to override individual block styles.
+.components-placeholder.components-placeholder {
+	// This needs specificity to override individual block styles.
```
I believe the comment here is in the wrong place. Maybe try moving it above the selector?

ciampo replied:
Good point. I believe this change happened with auto-formatting when saving the file.

Addressed in https://github.com/WordPress/gutenberg/pull/31713/commits/030c1bfeab295f402867a3a419e624855f367e0c

kevin940726 reviewed (commented):
I like this a lot!

ciampo committed:
Use `expect( console ).toHaveWarned()` instead of mocking `@wordpress/deprecated`, move new tests into one file

ciampo committed:
Revert auto-formatting

ciampo committed:
Move SCSS comment above selector

ciampo committed:
Do not mark isPrimary, isSecondary, isTertiary and isLink props as deprecated

ciampo commented:
Rebased on top of latest `trunk`, and removed the official deprecation notices in https://github.com/WordPress/gutenberg/pull/31713/commits/8344c800528e1dd1eb073c3e629518b4e7a0d50d.

The [docs](https://github.com/WordPress/gutenberg/pull/31713/files#diff-67ab236c71ce64ad2b69b520e436360298871296da9edc93901033e9519193bf) don't mention the `isPrimary`, `isSecondary`, `isTertiary` and `isLink` props anymore, in favor of the new `variant` prop

sarayourfriend reviewed (approved):
Once this is rebased it LGTM, nice work.

@youknowriad any ideas when we _would_ be able to actually add deprecation warnings for the old props?

youknowriad commented:
Right after 5.8 beta freeze would work for me. I guess that means next week :)

ciampo committed:
Remove TODO comment

ciampo committed:
Better wording for the notice component README

ciampo commented on packages/edit-post/src/components/header/template-title/delete-template.js:
```diff
@@ -46,8 +46,7 @@ export default function DeleteTemplate() {
 				isDestructive
-				isTertiary
-				isLink
+				variant="tertiary"
```
Hey @jorgefilipecosta ,  we are currently in the process of replacing `isTertiary` and `isLink` with a new `variant` prop — which also means that a button can be assigned to only one variant at the time.

Does this fit your needs? Or did you need this button to be at the same tertiary AND marked as a link?
We're also considering keeping `isLink` (see https://github.com/WordPress/gutenberg/pull/31713#discussion_r636480416)

jorgefilipecosta replied:
The design is changed a little bit. I think it is ok (provided we correct a small pixel alignment between the button and the input as a follow up):
new:
![detele button new](https://user-images.githubusercontent.com/11271197/119343957-ff53ec00-bc8e-11eb-84c2-55ec2eb3e253.jpg)


old:
![button-old](https://user-images.githubusercontent.com/11271197/119343983-05e26380-bc8f-11eb-825c-78f421b51e64.jpg)


But I'm going to cc: @jameskoster to double-check if we can do this UI change.

ciampo replied:
I could also change it to `variant="link"` if that works better

jameskoster replied:
Should be fine, I actually have a [PR open](https://github.com/WordPress/gutenberg/pull/32042) to adjust that popover a bit. There it's only using `isSecondary` and `isDestructive` which I assume will work fine with the new system?

ciampo replied:
> There it's only using isSecondary and isDestructive which I assume will work fine with the new system?

That's not going to be an issue. `isSecondary` will become `variant="secondary"` and `isDestructive` can stay as-is

mtias commented on packages/block-directory/src/plugins/get-install-missing/index.js:
```diff
@@ -88,7 +88,7 @@ const ModifiedWarning = ( { originalBlock, ...props } ) => {
 		);
 		actions.push(
-			<Button key="convert" onClick={ convertToHTML } isLink>
+			<Button key="convert" onClick={ convertToHTML } variant="link">
```
Wonder if it's slightly odd to see `link` as a variant. Would it ever make sense to have a "primary link" or a "secondary link"? Should `externalLink` be considered a variant as well (it exists as a separate thin now)?

sarayourfriend replied:
`variant="link"` doesn't determine whether the component renders an `anchor` tag, that's determined by the `href` prop's presence.

Furthermore, the link variant cannot be mixed with the other three variants for a good looking experience. These are the three variants together with `link`:

<img width="113" alt="Captura de Tela 2021-05-21 às 06 48 34" src="https://user-images.githubusercontent.com/24264157/119147830-cbfc2d80-ba00-11eb-8072-c573f5716d25.png">
<img width="114" alt="Captura de Tela 2021-05-21 às 06 48 22" src="https://user-images.githubusercontent.com/24264157/119147832-cc94c400-ba00-11eb-9dd6-81c744846442.png">
<img width="109" alt="Captura de Tela 2021-05-21 às 06 48 04" src="https://user-images.githubusercontent.com/24264157/119147835-cc94c400-ba00-11eb-8d13-babeec4b715a.png">

They all say "tertiary" button but they're the mixes of link with tertiary, secondary and primary respectively.

mtias replied:
If it's determined by `href` why do we need it as a variant?

sarayourfriend replied:
`link` is a variant of the _style_ of the button, not the functionality. Passing `href` changes the functionality. You could, for example, have a `button` that looks like a link by passing `<Button onClick={() => alert('hello!')} variant="link">Hello!</Button>`

kevin940726 replied:
We should probably rename it to `text` or `flat` though. [`material-ui`](https://material-ui.com/components/buttons/) is a good source of inspiration for this. (not necessary related to this PR though)

sarayourfriend replied:
I'd expect `text` to look most like the variantless button though, i.e., without the underline. 🤷‍♀️ I'm fine renaming it to whatever makes the most sense to people but I don't think we'll find something perfect. Our variants aren't as "varied" as material's. We really just have the five: `no variant`, `primary`, `secondary`, `tertiary`, and `link`. And they all or remove the outline/background arbitrarily (`link`, incidentally, is probably the easiest to guess what it does currently aside from the confusion around functionality).

lynnbgriffith replied:
I agree. I think it would be best to merge this change (i.e. introducing variant as a preferred alternative to the overlapping `is` props) and then we could potentially reconsider renaming or changing variants separately. Don't want to keep this PR open any longer because it touches so many usages of Button that it needs constant rebasing, etc.

mtias replied:
It's weird to use "link" to refer to an appearance characteristic but we can revisit this separately as we should also clarify the design expectations.

mtias commented:
I like this.

> Do not mark the isPrimary, isSecondary, isTertiary and isLink props as deprecated yet, but add logic to the <Button /> component to translate their values into the variant prop

This sounds like a good approach to me. I think we might consider soft deprecating — just converting internally but not warning — given how widespread the use of this component is.

Lovor01 commented:
This change is introduced in the docs too early. WordPress 5.8.1 still uses button with is**** attributes, and these deprecations (will still work in future versions - I inspected code) are not mentioned at all in the docs for Button component

ciampo commented:
Hey @Lovor01!

> This change is introduced in the docs too early [...] these deprecations  are not mentioned at all in the docs for Button component

This deprecation was mentioned [in the changelog](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md#1420-2021-07-21).

Regarding the mention on the `Button` docs, we are on purpose not mentioning it. We just removed the props from the docs (while still supporting them in code), in order to encourage the usage of the new `variant` prop without introducing a breaking change. You can read more about this approach in the [contributing guidelines](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#compatibility)

> WordPress 5.8.1 still uses button with is**** attributes

Could you be more precise about which Buttons in the codebase you are referring to?
